### PR TITLE
#94 rewrite RefreshIDToken middleware to use prompt=none

### PR DESCRIPTION
**Pull request by willkg**
This reworks the `RefreshIDToken` middleware to use oidc authorize endpoint with prompt=none.

I also moved all of it from contrib/auth0/ into the root of the project since it's no longer auth0 specific.

I updated docs and tests.

### CODE REVIEW

**Comment by willkg** [comment]
I'm pretty sure I got this right. There's some duplication between the view and the middleware, but it's minor and it's unlikely we'll use it in a third place.

I tested it with Standup. One thing I noted is that it's impossible to tell if it's working right. I think I want to add `logging.debug` calls in a bunch of places so we can more easily trace it when verifying things are working.

Does that sound ok? We don't seem to do that anywhere else in the library, but I think we should.

File: mozilla_django_oidc/middleware.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+# FIXME(willkg): This doesn't appear to be used anywhere.
+def logout_url():
+    """Log out the user from Auth0."""
+    url = 'https://' + import_from_settings('OIDC_OP_DOMAIN') + '/v2/logout'
+    url += '?' + urlencode({
+        'returnTo': import_from_settings('LOGOUT_REDIRECT_URL', '/'),
+        'client_id': import_from_settings('OIDC_RP_CLIENT_ID')
+    })
+    return url
```
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: This doesn't appear to be used anywhere so I added a FIXME to it. If it turns out it is used, then I can remove it.

File: mozilla_django_oidc/middleware.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+class RefreshIDToken(MiddlewareMixin):
+    """Renews id_tokens after expiry seconds
+
+    For users authenticated with an id_token, we need to check that it's still
+    valid after a specific amount of time and if not, force them to
+    re-authenticate silently.
+
+    """
+    def process_request(self, request):
+        if ((request.method == 'GET' and
```
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: Because the middleware now redirects to the authorization endpoint if the token has expired, pretty sure we need to restrict it to just GET requests.

Does that sound right? Are there other methods that will work with the redirects?

File: mozilla_django_oidc/middleware.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+             request.user.is_authenticated() and
+             not request.path.startswith('/oidc/') and
```
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: I added this because we don't want it to kick off during OIDC related requests like `/oidc/callback`.

File: mozilla_django_oidc/middleware.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+             not request.is_ajax())):
+            expiration = request.session.get('oidc_id_token_expiration')
+            if expiration is not None and expiration > time.time():
```
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: `time.time()` will give back seconds since epoch for that server. If all the servers in a cluster are in the same timezone, that'll work fine. If they're not then this won't work.

What's a better way to do this? Should we be using `YYYY-mm-dd HH:MM:SS` UTC timestamps and string comparisons?

File: mozilla_django_oidc/middleware.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                    'nonce': nonce
+                })
+                request.session['oidc_nonce'] = nonce
+
+            request.session['oidc_state'] = state
+            request.session['oidc_login_next'] = request.get_full_path()
+
+            query = urlencode(params)
+            redirect_url = '{url}?{query}'.format(url=auth_url, query=query)
+            return HttpResponseRedirect(redirect_url)
```
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: This method is the meat of the change. If the id token has expired, then this figures out the authentication endpoint url to redirect to and sets up the session and other bits accordingly. Then it redirects and the SSO will redirect back to the callback url if all goes well.

File: mozilla_django_oidc/middleware.py (hunk 4 above)
Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: I don't get this. The epoch is defined with a timezone. It was UTC. 
If I generate a `time.time()` on my server in New York, I get the same number as a server in London. Even though they have different values in `/etc/timezone`.

Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: Is it worth saving some characters by doing:
```python
expiration = request.session.get('oidc_id_token_expiration', 0)
if expiration > time.time():
```
??

Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: I wonder if this line should be put first in this function. If the expiration is set and it's "future" then there's no point waking up `request.user` or `request.is_ajax` etc.

File: mozilla_django_oidc/middleware.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            auth_url = import_from_settings('OIDC_OP_AUTHORIZATION_ENDPOINT')
+            client_id = import_from_settings('OIDC_RP_CLIENT_ID')
+            state = get_random_string(import_from_settings('OIDC_STATE_SIZE', 32))
+
+            # Build the parameters as if we were doing a real auth handoff, except
+            # we also include prompt=none.
+            params = {
+                'response_type': 'code',
+                'client_id': client_id,
+                'redirect_uri': absolutify(reverse('oidc_authentication_callback')),
```
Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: Perhaps there's something I don't "get" about this `absolutify`. It depends on a config setting which you have to maintain for dev, stage and prod. To accomplish the same result you can just use the request. E.g.

```python
    'redirect_uri': request.build_absolute_uri(reverse('oidc_authentication_callback')),
```

File: mozilla_django_oidc/middleware.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+            if import_from_settings('OIDC_USE_NONCE', True):
+                nonce = get_random_string(import_from_settings('OIDC_NONCE_SIZE', 32))
+                params.update({
+                    'nonce': nonce
+                })
+                request.session['oidc_nonce'] = nonce
+
+            request.session['oidc_state'] = state
+            request.session['oidc_login_next'] = request.get_full_path()
```
Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: Neat!

File: mozilla_django_oidc/utils.py [hunk 8]
```diff
@@ -2,7 +2,6 @@
     from urlparse import urljoin
 except ImportError:
     from urllib.parse import urljoin
-
 from django.conf import settings
```
Comment (peterbe) [inline mozilla_django_oidc/utils.py]: That break was there to separate standard lib from third-party.

File: mozilla_django_oidc/middleware.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+from django.core.urlresolvers import reverse
+from django.http import HttpResponseRedirect
+from django.utils.crypto import get_random_string
+
+from mozilla_django_oidc.utils import absolutify, import_from_settings
+
+
+# Django 1.11 makes changes to how middleware work. In Django 1.11+, we want to
+# use the mixin so that our middleware works as is.
+if tuple(django.VERSION[0:2]) >= (1, 10):
```
Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: First of all, `django.VERSION[0:2]` is already a tuple. And you don't need the slice. 
This is the exact same meaning: `if django.VERSION >= (1, 10):` and it's neater. 

Secondly, the *comment* above says Django 1.11 but this if statement is about Django 1.10.

Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: I suspect this is unrelated to the PR at hand. You can ignore this comment and we can deal with it separately.

File: mozilla_django_oidc/middleware.py (hunk 4 above)
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: Ok--sounds like this is fine, then.

File: mozilla_django_oidc/middleware.py (hunk 6 above)
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: I'm mimicking the code in the authentication view.

File: mozilla_django_oidc/utils.py (hunk 8 above)
Comment (willkg) [inline mozilla_django_oidc/utils.py]: Oops--that shouldn't have been in the PR.

**Comment by willkg** [comment]
> The more I look at this the more I wonder how this will ever work with a modern web app where the Django server just spits out RESTy JSON to a single-page app.

I implemented what @gdestuynder suggested in #94 as best as I could figure. If I didn't do it right, let me know. I have no idea how this works with AJAXy web-apps, but as far as I can see, the existing middleware doesn't work with those, either.

File: mozilla_django_oidc/middleware.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        pass
+
+
+OIDC_URLS = [
+    reverse(name)
+    for name in [
+        'oidc_authentication_init',
+        'oidc_authentication_callback',
+        'oidc_logout'
+    ]
```
Comment (johngian) [inline mozilla_django_oidc/middleware.py]: nit: I think that a config entry to allow devs to extend this list with custom URLs might be handy

File: mozilla_django_oidc/middleware.py (hunk 2 above)
Comment (johngian) [inline mozilla_django_oidc/middleware.py]: Does it make sense to add a catch-all case were we force users to log out ?

File: mozilla_django_oidc/middleware.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+                return
+
+            # The id_token has expired, so we have to re-authenticate silently.
+            auth_url = import_from_settings('OIDC_OP_AUTHORIZATION_ENDPOINT')
+            client_id = import_from_settings('OIDC_RP_CLIENT_ID')
+            state = get_random_string(import_from_settings('OIDC_STATE_SIZE', 32))
+
+            # Build the parameters as if we were doing a real auth handoff, except
+            # we also include prompt=none.
+            params = {
```
Comment (johngian) [inline mozilla_django_oidc/middleware.py]: nit: Since we are copying the code from the authentication request, should we write a helper function and call it accordingly?

File: mozilla_django_oidc/middleware.py (hunk 10 above)
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: How would that work?

Would we have it as a setting and have it default to these three? That seems hard to use for developers.

Would we have a setting and always append these three?

File: mozilla_django_oidc/middleware.py (hunk 11 above)
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: I mentioned that in the first PR comment. Namely, there is some duplication, but it'll only be in two places and they have subtle differences, so I'm not sure it helps a lot to centralize it into a helper function.

File: mozilla_django_oidc/middleware.py (hunk 2 above)
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: This forces a re-authentication in a bunch of cases which will result in the user getting logged out. I covered all the cases I could think of, but if there are others, we should make sure they're covered, too.

File: mozilla_django_oidc/middleware.py (hunk 10 above)
Comment (johngian) [inline mozilla_django_oidc/middleware.py]: I was thinking the latter. Having a config entry that defaults to an empty list and we always append these 3 entries.

**Comment by willkg** [comment]
^^^ That rebases against master tip to pick up @peterbe's `is_authenticated` changes. It also fixes nits @peterbe had.

I want to add logging to this because otherwise it's impossible to see if it's working.

I also want someone who knows OIDC really well and what @gdestuynder wanted in #94 to make sure the changes I made are right.

I'll work on adding debug-level logging now.

File: mozilla_django_oidc/middleware.py (hunk 10 above)
Comment (willkg) [inline mozilla_django_oidc/middleware.py]: That makes sense--I'll add that.

File: mozilla_django_oidc/middleware.py (hunk 6 above)
Comment (peterbe) [inline mozilla_django_oidc/middleware.py]: Once this lands, I'll bring it up again. I might need a refresher as to why or why not you can't use the domain that is in the `request` object ...from a security point of view.

**Comment by willkg** [comment]
^ That redoes the exempt urls code and also makes the middleware more flexible in regards to what constitutes a refreshable url.

I also added some light logging which makes it much easier to verify that the middleware is doing something.

I also added some troubleshooting docs.

@johngian Does this look ok? Is the implementation for the `prompt=none` part correct?

File: mozilla_django_oidc/middleware.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+class RefreshIDToken(MiddlewareMixin):
+    """Renews id_tokens after expiry seconds
+
+    For users authenticated with an id_token, we need to check that it's still
+    valid after a specific amount of time and if not, force them to
+    re-authenticate silently.
+
+    """
+    def get_exempt_urls(self):
+        """Generate a return a set of url paths to exempt from RefreshIDToken
```
Comment (johngian) [inline mozilla_django_oidc/middleware.py]: typo: Generate and return

**Comment by johngian** [comment]
Code looks OK. Please squash and merge 🚀

**Comment by willkg** [comment]
^^^ That fixes the typo. After that passes, I'll squash and merge.

Thank you!

**Comment by gdestuynder** [comment]
looks alright to me (went through the code quickly)
